model/riscv_stateen_regs.sail

```diff
@@ -0,0 +1,484 @@
+    2 => 0xFFFFFFFF @ 0x00000000,
+    3 => 0xFFFFFFFF @ 0x00000000,
+  }
+}
+
+mapping stateen_idx : {0, 1, 2, 3} <-> bits(4) = {
+  0 <-> 0b1100, // 0xC
+  1 <-> 0b1101, // 0xD
+  2 <-> 0b1110, // 0xE
+  3 <-> 0b1111, // 0xF
+}
+
+enum stateen_control_bits = {STATEEN_FCSR, STATEEN_ENVCFG, STATEEN_SE}
+
+mapping csr_permission_idx : stateen_control_bits <-> {1, 62, 63} = {
+  STATEEN_SE     <-> 63,
+  /* stateen0 */
+  STATEEN_ENVCFG <-> 62,
+  STATEEN_FCSR   <-> 1,
+}
+
+/* Computes a bit-level permission mask by performing a bitwise AND across the
+ * stateen bits from all higher privilege modes. Access to a corresponding bitfield
+ * is granted only if each bit is set to 1 in the same bitfield of the stateen
+ * registers at all higher privilege modes.
+ */
+function get_stateen_permission_mask (idx : {0, 1, 2, 3}) -> bits(64) = {
+  let mstateen = get_mstateen(idx);
+  let hstateen = get_hstateen(idx);
+  let sstateen = get_sstateen(idx);
+  match cur_privilege {
+    Machine    => 0xFFFFFFFFFFFFFFFF,
+    Supervisor => mstateen & hstateen,
+    User       => mstateen & hstateen & sstateen,
+  }
+}
+
+/*
+ * Maps a CSR address to the corresponding stateen permission bit and stateen id
+ *
+ * Returns an optional tuple:
+ *   (stateen bitfield, stateen id)
+ *
+ * Add more mappings as needed
+ */
+function csr_to_stateen_bit(csr : csreg) -> option((stateen_control_bits, {0, 1, 2, 3})) = {
+  match csr {
+    /* stateen0 */
+    0x10C /* sstateen0 */ => Some((STATEEN_SE, 0)),
+    0x10A /* senvcfg   */ => Some((STATEEN_ENVCFG, 0)),
+    0x001 /* fflags    */ => if currentlyEnabled(Ext_Zfinx) then Some((STATEEN_FCSR, 0)) else None(),
+    0x002 /* frm       */ => if currentlyEnabled(Ext_Zfinx) then Some((STATEEN_FCSR, 0)) else None(),
+    0x003 /* fcsr      */ => if currentlyEnabled(Ext_Zfinx) then Some((STATEEN_FCSR, 0)) else None(),
+
+    /* stateen1 */
+    0x10D /* sstateen1 */ => Some((STATEEN_SE, 1)),
+
+    /* stateen2 */
+    0x10E /* sstateen2 */ => Some((STATEEN_SE, 2)),
+
+    /* stateen3 */
+    0x10F /* sstateen3 */ => Some((STATEEN_SE, 3)),
+
+    _ => None(),
+  }
+}
+
+/*
+ * Verifies whether access to CSRs protected by the Smstateen or Ssstateen
+ * extensions is allowed at the current privilege level.
+ */
+function check_CSR_stateen_permission(csr : csreg) -> bool = {
+  match csr_to_stateen_bit(csr) {
+    Some((bit_idx, stateen_register_index)) => {
+      let mask = get_stateen_permission_mask(stateen_register_index);
+      mask[csr_permission_idx(bit_idx)] == bitone
+    },
+    None() => true
+  }
+}
+
+/*
+ * Check if floating-point instructions using X registers
+ * are permitted in the given privilege mode.
+ *
+ * Note: Machine mode can always perform floating-point instructions
+ */
+function check_fp_stateen() -> bool = {
+  let mask = get_stateen_permission_mask(0);
+  mask[csr_permission_idx(STATEEN_FCSR)] == bitone
+}
+
+function reset_stateen() -> unit = {
+  if enable_mstateen() then {
```

Comment:
There's no need for this check - if it isn't enabled then there's no harm resetting them.